File: plugins/android/weiui_pay/src/main/java/cc/weiui/pay/weiui/ui/module/weiuiPayModule.java
package cc.weiui.pay.weiui.ui.module;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.os.Handler;
import android.os.Message;

import com.alipay.sdk.app.PayTask;
import com.chinaums.pppay.unify.UnifyPayListener;
import com.chinaums.pppay.unify.UnifyPayPlugin;
import com.chinaums.pppay.unify.UnifyPayRequest;
import com.taobao.weex.annotation.JSMethod;
import com.taobao.weex.bridge.JSCallback;
import com.taobao.weex.common.WXModule;
import com.tencent.mm.opensdk.modelbase.BaseResp;
import com.tencent.mm.opensdk.modelpay.PayReq;
import com.tencent.mm.opensdk.openapi.IWXAPI;
import com.tencent.mm.opensdk.openapi.WXAPIFactory;

import java.util.HashMap;
import java.util.Map;

import cc.weiui.framework.extend.module.weiuiCommon;
import cc.weiui.framework.extend.module.weiuiJson;
import cc.weiui.pay.library.alipay.PayResult;
import cc.weiui.pay.library.weixin.PayStatic;

public class weiuiPayModule extends WXModule {

    private static final String TAG = "weiuiPayModule";

    private static final int SDK_PAY_FLAG = 1;

    private Map<String, JSCallback> msgJSCallback = new HashMap<>();

    @SuppressLint("HandlerLeak")
    @SuppressWarnings("unchecked")
    private Handler mHandler = new Handler() {
        public void handleMessage(Message msg) {
            switch (msg.what) {
                /**
                 * 支付宝结果回调
                 */
                case SDK_PAY_FLAG:
                    PayResult payResult = new PayResult((Map<String, String>) msg.obj);
                    JSCallback mJSCallback = msgJSCallback.get(payResult.getMsgName());
                    if (mJSCallback != null) {
                        Map<String, Object> data = new HashMap<>();
                        data.put("status", payResult.getResultStatus());
                        data.put("result", payResult.getResult());
                        data.put("memo", payResult.getMemo());
                        mJSCallback.invoke(data);
                        msgJSCallback.remove(payResult.getMsgName());
                        break;
                    }

                default:
                    break;
            }
        }
    };

    /**
     * 微信支付结果回调
     * @param resp
     */
    public static void onWeixinResp(BaseResp resp) {
        if (PayStatic.payCallback != null) {
            Map<String, Object> data = new HashMap<>();
            data.put("status", resp.errCode);
            if (resp.errCode == 0) {
                data.put("msg", "支付成功");
            } else if (resp.errCode == -1) {
                data.put("msg", "可能的原因：签名错误、未注册APPID、项目设置APPID不正确、注册的APPID与设置的不匹配、其他异常等");
            } else if (resp.errCode == -2) {
                data.put("msg", "用户取消");
            } else {
                return;
            }
            PayStatic.payCallback.invoke(data);
            PayStatic.payCallback = null;
        }
    }

    /***************************************************************************************************/
    /***************************************************************************************************/
    /***************************************************************************************************/

    /**
     * 官方微信支付
     *
     * @param payData
     * @param callback
     */
    @JSMethod
    public void weixin(String payData, JSCallback callback) {
        PayStatic.payParamets = weiuiJson.parseObject(payData);
        PayStatic.payCallback = callback;
        IWXAPI api = WXAPIFactory.createWXAPI(mWXSDKInstance.getContext(), null);
        api.registerApp(weiuiJson.getString(PayStatic.payParamets, "appid"));
        PayReq request = new PayReq();
        request.appId = weiuiJson.getString(PayStatic.payParamets, "appid");
        request.partnerId = weiuiJson.getString(PayStatic.payParamets, "partnerid");
        request.prepayId = weiuiJson.getString(PayStatic.payParamets, "prepayid");
        request.packageValue = weiuiJson.getString(PayStatic.payParamets, "package");
        request.nonceStr = weiuiJson.getString(PayStatic.payParamets, "noncestr");
        request.timeStamp = weiuiJson.getString(PayStatic.payParamets, "timestamp");
        request.sign = weiuiJson.getString(PayStatic.payParamets, "sign");
        if (!api.sendReq(request)) {
            if (PayStatic.payCallback != null) {
                Map<String, Object> data = new HashMap<>();
                data.put("status", -999);
                data.put("msg", "启动微信支付失败");
                PayStatic.payCallback.invoke(data);
                PayStatic.payCallback = null;
            }
        }
    }

    /**
     * 官方支付宝支付
     *
     * @param payData
     * @param callback
     */
    @JSMethod
    public void alipay(String payData, JSCallback callback) {
        final String orderInfo = payData;
        final String msgName = "JSCallback_" + weiuiCommon.randomString(6);
        msgJSCallback.put(msgName, callback);
        Runnable payRunnable = new Runnable() {

            @Override
            public void run() {
                PayTask alipay = new PayTask((Activity) mWXSDKInstance.getContext());
                Map<String, String> result = alipay.payV2(orderInfo, true);
                result.put("msgName", msgName);

                Message msg = new Message();
                msg.what = SDK_PAY_FLAG;
                msg.obj = result;
                mHandler.sendMessage(msg);
            }
        };
        Thread payThread = new Thread(payRunnable);
        payThread.start();
    }

    /**
     * 银联微信支付（无回调功能）
     *
     * @param payData
     */
    @JSMethod
    public void union_weixin(String payData) {
        UnifyPayRequest payRequest = new UnifyPayRequest();
        payRequest.payChannel = UnifyPayRequest.CHANNEL_WEIXIN;
        payRequest.payData = payData;
        UnifyPayPlugin.getInstance(mWXSDKInstance.getContext()).setListener(new UnifyPayListener() {
            @Override
            public void onResult(String resultCode, String resultInfo) {

            }
        }).sendPayRequest(payRequest);
    }

    /**
     * 银联支付宝支付（无回调功能）
     *
     * @param payData
     */
    @JSMethod
    public void union_alipay(String payData) {
        UnifyPayRequest payRequest = new UnifyPayRequest();
        payRequest.payChannel = UnifyPayRequest.CHANNEL_ALIPAY;
        payRequest.payData = payData;
        UnifyPayPlugin.getInstance(mWXSDKInstance.getContext()).setListener(new UnifyPayListener() {
            @Override
            public void onResult(String resultCode, String resultInfo) {

            }
        }).sendPayRequest(payRequest);
    }
}

File: platforms/ios/WeexWeiui/WeexWeiui/Weiui/Moduld/WeiuiPayModule.m

#import "WeiuiPayModule.h"
#import <AlipaySDK/AlipaySDK.h>
#import "WXApi.h"
#import "UMSPPPayUnifyPayPlugin.h"

static WXModuleKeepAliveCallback alipayCallback;
static WXModuleKeepAliveCallback weixinCallback;

@implementation WeiuiPayModule

WX_EXPORT_METHOD(@selector(weixin:callback:))
WX_EXPORT_METHOD(@selector(alipay:callback:))
WX_EXPORT_METHOD(@selector(union_weixin:))
WX_EXPORT_METHOD(@selector(union_alipay:))

+ (void)alipayHandleOpenURL:(NSURL *) url
{
    [[AlipaySDK defaultService] processOrderWithPaymentResult:url standbyCallback:^(NSDictionary *resultDic) {
        [WeiuiPayModule onAlipayResp:resultDic];
    }];
}

+ (BOOL)weixinHandleOpenURL:(NSURL *) url
{
    return [WXApi handleOpenURL:url delegate:(id<WXApiDelegate>)[[WeiuiPayModule alloc] init]];
}

//支付宝结果回调
+ (void)onAlipayResp:(NSDictionary *)result
{
    NSMutableDictionary *data = [[NSMutableDictionary alloc] init];
    if (result != nil) {
        [data setObject:result[@"resultStatus"] forKey:@"status"];
        [data setObject:result[@"result"] forKey:@"result"];
        if ([data[@"status"] isEqualToString:@"9000"] && !result[@"memo"]) {
            [data setObject:@"支付成功" forKey:@"memo"];
        }else{
            [data setObject:result[@"memo"] forKey:@"memo"];
        }
    }
    if (alipayCallback != nil) {
        alipayCallback(data, NO);
        alipayCallback = nil;
    }
}

// 微信支付结果回调
- (void)onResp:(BaseResp *)resp
{
    if (weixinCallback != nil) {
        NSString *msg;
        if (resp.errCode == 0) {
            msg = @"支付成功";
        }else if (resp.errCode == -1) {
            msg = @"可能的原因：签名错误、未注册APPID、项目设置APPID不正确、注册的APPID与设置的不匹配、其他异常等";
        }else if (resp.errCode == -2) {
            msg = @"用户取消";
        }else{
            return;
        }
        weixinCallback(@{@"status":@(resp.errCode), @"msg":msg}, NO);
        weixinCallback = nil;
    }
}

/*******************************************************************************************/
/*******************************************************************************************/
/*******************************************************************************************/

//官方微信支付
- (void)weixin:(NSDictionary *)payData callback:(WXModuleKeepAliveCallback)callback
{
    weixinCallback = callback;
    [WXApi registerApp:payData[@"appid"]];
    PayReq *req  = [[PayReq alloc] init];
    req.partnerId = payData[@"partnerid"];
    req.prepayId = payData[@"prepayid"];
    req.package = payData[@"package"];
    req.nonceStr = payData[@"noncestr"];
    req.timeStamp = [payData[@"timestamp"] intValue];
    req.sign = payData[@"sign"];
    if (![WXApi sendReq:req]) {
        if (weixinCallback != nil) {
            weixinCallback(@{@"status":@(-999), @"msg":@"启动微信支付失败"}, NO);
            weixinCallback = nil;
        }
    }
}

//官方支付宝支付
- (void)alipay:(NSString*)payData callback:(WXModuleKeepAliveCallback)callback
{
    alipayCallback = callback;
    [[AlipaySDK defaultService] payOrder:payData fromScheme:@"weiuiApp_xxxxxxxx" callback:^(NSDictionary *resultDic) {
        [WeiuiPayModule onAlipayResp:resultDic];
    }];
}

//银联微信支付（无回调功能）
- (void)union_weixin:(NSString*)payData
{
    [UMSPPPayUnifyPayPlugin payWithPayChannel:CHANNEL_WEIXIN payData:payData callbackBlock:^(NSString *resultCode, NSString *resultInfo) {
        
    }];
}

//银联支付宝支付（无回调功能）
- (void)union_alipay:(NSString*)payData
{
    [UMSPPPayUnifyPayPlugin payWithPayChannel:CHANNEL_ALIPAY payData:payData callbackBlock:^(NSString *resultCode, NSString *resultInfo) {
        
    }];
}

@end
